src/Expander.hs

```diff
@@ -1301,22 +1311,28 @@ expandOneForm prob stx
           implV <- Env.lookupVal transformerName <$> currentTransformerEnv
           case implV of
             Just (ValueClosure macroImpl) -> do
-              macroVal <- inEarlierPhase $ expandEval $
-                          apply macroImpl $
-                          ValueSyntax $ addScope p stepScope stx
+              macroVal <- inEarlierPhase
+                          $ pure
```

Comment:
as discussed, `inEarlierPhase` uses `local` to change the current phase within the body, but `pure` can't access the current phase. is this a bug? or did the previous code simply didn't use the current phase, so this call to `inEarlierPhase` was already dead code?

Reply:
The `inEarlierPhase` was _not_ dead code. On the `main` branch, if I run this program:
```
#lang "prelude.kl"
(import (shift "prelude.kl" 1))

(define-macros
  ([my-error
    (lambda (stx)
      (error '"oops"))]))
(example (my-error))
```
Then I get the expected error message, namely that `error` was raised at phase 1, while evaluating the body of the macro:
```
Error at phase p1: <loc>: "oops"
```
But if I delete `inEarlierPhase` from this `EUserMacro` case in `main`, the error message instead claims that `error` was raised in phase 0, which is not correct:
```
Error at phase p0: <loc>: "oops"
``` 

When I run the above program in this `doyougnu/wip/useful-repl` branch, I instead get the error message
```
Not a macro monad value: erroneusValue: likely
```
Which I think says that that a primitive claims it has type `Macro something` but is returning something else, a string for example. Technically `(error '"oops")` shouldn't be returning anything at all though, it should have the side-effect of terminating the program early!

Reply:
thanks for introducing this bug, this allowed me to find a related bug in the code which immediately follows! the `interpretMacroAction` call should also be `inEarlierPhase`, but isn't.

it is surprising how little of an impact these two bugs have!

Reply:
my mistake, whether `inEarlierPhase` is or isn't added to the `interpretMacroAction` call does make a big difference.

adding `inEarlierPhase` makes so many tests fail that it's hard to tell which of the many error messages is the root cause. One relatively simple piece which breaks is `free-identifier=?`:
```
#lang "prelude.kl"
(import (shift "prelude.kl" 1))

(define-macros
  ([my-keyword1
    (lambda (stx)
      (syntax-error '"my-keyword1 is a keyword"))]))
(define my-keyword2 "my-keyword2 is a keyword")
(define-macros
  ([my-macro
    (lambda (stx)
      (case (open-syntax stx)
        [(list-contents (:: _ (:: x (nil))))
         (>>= (free-identifier=? x 'my-keyword1)
           (lambda (r)
             (if r
               (pure '"input is my-keyword1")
               (>>= (free-identifier=? x 'my-keyword2)
                 (lambda (r)
                   (if r
                     (pure '"input is my-keyword2")
                     (pure '"input is not a keyword")))))))]))]))
(example (my-macro my-keyword1))
(example (my-macro my-keyword2))
```
on `main` the above program prints "input is my-keyword1" and "input is my-keyword2", whereas with `inEarlierPhase $ interpretMacroAction prob act` the program instead prints "input is not a keyword" twice.

Reply:
here's my reasoning for why I think `inEarlierPhase` is needed in both places. Let's look at another simple Klister program.
```
#lang "prelude.kl"
(import (shift "prelude.kl" 1))

(define x 42)
(meta
  (define >>
    (lambda (action1 action2) 
      (>>= action1
        (lambda (_)
          action2)))))
(define-macros
  ([my-macro
    (lambda (stx)
      (>> (log '"inside my-macro")
          (pure 'x)))]))
(example (my-macro))
```

In order to run this program (at phase 0), we must first expand it into Core. When expanding `(my-macro)`, the expander notices that this is a user-defined macro (the `EUserMacro transformerName -> ...` case is selected) and the definition of `my-macro` is looked up (`Env.lookupVal`). It is looked up in the phase 0 _transformer_ environment, which I like to think as being between the phase 0 _regular_ environment and the phase 1 regular environment.

We find the definition:
```
(lambda (stx)
  (>> (log '"inside my-macro")
    (pure 'x)))
```
This is a pure expression whose variables, such as `>>`, come from the phase 1 regular environment (and whose macros come from the phase 1 transformer environment). We must thus evaluate the application of this phase 1 expression applied to the syntax object `'(my-macro)` in phase 1, hence 
```
inEarlierPhase $ expandEval $
  apply macroImpl $
  ValueSyntax {- (my-macro) -}
```

The result of this evaluation is a `Macro Syntax` action. We're still inside the macro definition, so I would expect to execute this action inside phase 1, hence `inEarlierPhase $ interpretMacroAction prob act`.

But `my-keyword1` and `my-keyword2` live in the phase 0 transformer environment and in the phase 0 regular environment, respectively. Is that why the macro action needs to run in phase 0? That still seems wrong to me. Yes, `my-keyword1` and `my-keyword2` live in phase 0, but they are _quoted_ identifiers, so I would expect `my-keyword2` to be a phase 0 expression and `'keyword2` to be a phase 1 expression referring to a phase 0 expression!

Is it possible that we _do_ need `inEarlierPhase $ interpretMacroAction prob act`, as I expect, and that it is `free-identifier=?` which is wrong, because it looks in the current environment but should be looking in the environment for the next phase? Or is it me whose understanding is completely backwards?

Reply:
I created #240 to verify my understanding of the bug independently of this MR. but we should not merge #240, as it would just conflict with and get overwritten by this MR. Instead, the real fix must be performed on the CEK evaluator. I think I will attempt to write that fix, that will force me to understand, and thus review, the CEK evaluator!

Reply:
turns out the CEK evaluator does not need to be changed to address this bug, since the _evaluation_ does not depend on the current phase), only the _execution_ and _expansion_ does.